pufferlib/environments/ocean/connect4/connect4.h

```diff
@@ -128,104 +130,113 @@ uint64_t bottom_mask(uint64_t column) {
+        return true;
+    }
 
     // Vertical;
-    m = position & (position >> 1);
-    if(m & (m >> 2)) return true;
+    m = pieces & (pieces >> 1);
+    if(m & (m >> 2)) {
+        return true;
+    }
 
     return false;
 }
 
-float minmax(uint64_t position, uint64_t mask, int depth, bool maximising_player) {
-    // https://en.wikipedia.org/wiki/Minimax
-    bool has_won = won(position);
-    if (depth == 0 || has_won || draw(mask)) {
-        if (has_won) {
-            float discount = (4 - depth) * 0.1; // Discount short games
-            return maximising_player ? WIN_VALUE + discount : -WIN_VALUE - discount;
+float negamax(uint64_t pieces, uint64_t other_pieces, int depth) {
+    // https://en.wikipedia.org/wiki/Negamax#Negamax_variant_with_no_color_parameter
+    uint64_t piece_mask = mask(pieces, other_pieces);
+    
+    bool opp_won = won(other_pieces);
+    if (depth == 0 || opp_won || draw(piece_mask)) {
+        if (opp_won) {
+            return -WIN_VALUE;
         }
-        return DRAW_VALUE;
+        // Break ties between non-terminal states
+        return (float) (rand() % 10);
     }
 
-    float value = maximising_player ? -MAX_VALUE : MAX_VALUE;
+    float value = -MAX_VALUE;
     for (uint64_t column = 0; column < 7; column ++) {
-        if (!valid_move(column, mask)) {
+        if (invalid_move(column, piece_mask)) {
             continue;
         }
-        uint64_t child_position = swap_player(position, mask);
-        uint64_t child_mask = play(column, mask);
-        float child_value = minmax(child_position, child_mask, depth - 1, !maximising_player);
-        if (maximising_player && child_value > value) {
-            value = child_value;
-        }
-        if (!maximising_player && child_value < value) {
-            value = child_value;
-        }
+        uint64_t child_pieces = play(column, piece_mask, other_pieces);
+        value = fmax(value, -negamax(other_pieces, child_pieces, depth - 1));
     }
     return value;
 }
 
 int compute_env_move(CConnect4* env) {
-    uint64_t hash = env->position + env->mask + bottom();
-    // Hard coded opening book
+    uint64_t piece_mask = mask(env->player_pieces, env->env_pieces);
+    
+    // Compute a unique hash for this game state
+    uint64_t hash = env->player_pieces + piece_mask + bottom();
+
+    // Hard coded opening book to handle some early game traps
     // TODO: Add more opening book moves
-    if (hash == 4398048608256) {
-        // Respond to _ _ _ o _ _ _
-        // with       _ _ x o _ _ _
-        // Which is a common early game trap
-        return 2;
+    switch (hash) {
+        case 4398050705408:
+            // Respond to _ _ _ o _ _ _
+            // with       _ _ x o _ _ _
+            return 2;
+        case 4398583382016:
+            // Respond to _ _ _ _ o _ _
+            // with       _ _ _ x o _ _
+            return 3;
     }
 
+    // https://en.wikipedia.org/wiki/Negamax#Negamax_variant_with_no_color_parameter
```

Comment:
No need to double comment algo. I'd probably just put something like this at top of file for context... don't even really need to link, just name the algo